test/jdk/sun/net/www/http/KeepAliveCache/B8291637.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8291637
+ * @run main/othervm -Dhttp.keepAlive.time.server=20 -esa -ea B8291637 timeout
+ * @run main/othervm -Dhttp.keepAlive.time.server=20 -esa -ea B8291637 max
+ */
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.*;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.CompletableFuture;
+
+public class B8291637 {
+    static CompletableFuture<Boolean> passed = new CompletableFuture<>();
+
+    static class Server extends Thread {
+        final ServerSocket serverSocket;
+        final int port;
+        final String param; // the parameter to test "max" or "timeout"
+        volatile Socket s;
+
+        public Server(String param) throws IOException {
+            serverSocket = new ServerSocket(0);
+            port = serverSocket.getLocalPort();
+            setDaemon(true);
+            this.param = param;
+        }
+
+        public int getPort() {
+            return port;
+        }
+
+        public void close() {
+            try {
+                serverSocket.close();
+                s.close();
```

Comment:
Could that trigger a NPE if `s` is null?

Reply:
That could happen only if `ServerSocket.accept` throws an exception, which I guess is possible. But, the client side would not get to call `Server.close` then. I'll add a check anyway.